File: src/sunstone/public/app/opennebula/network.js
/* -------------------------------------------------------------------------- */
/* Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/* -------------------------------------------------------------------------- */

define(function(require) {
  var OpenNebulaAction = require('./action');

  var RESOURCE = "VNET";

  var STATES = {
    INIT            : 0,
    READY           : 1,
    LOCK_CREATE     : 2,
    LOCK_DELETE     : 3,
    LOCKED          : 4,
    DONE            : 5,
    ERROR           : 6
  };

  var STATES_STR = [
    "INIT",
    "READY",
    "LOCK_CREATE",
    "LOCK_DELETE",
    "LOCKED",
    "DONE",
    "ERROR"
  ];

  var Network =  {
    "resource": RESOURCE,
    "create": function(params) {
      OpenNebulaAction.create(params, RESOURCE);
    },
    "del": function(params) {
      OpenNebulaAction.del(params, RESOURCE);
    },
    "list": function(params) {
      OpenNebulaAction.list(params, RESOURCE);
    },
    "list_in_zone" : function(params) {
      OpenNebulaAction.list_in_zone(params, RESOURCE);
    },
    "show": function(params) {
      OpenNebulaAction.show(params, RESOURCE);
    },
    "chown" : function(params) {
      OpenNebulaAction.chown(params, RESOURCE);
    },
    "chgrp" : function(params) {
      OpenNebulaAction.chgrp(params, RESOURCE);
    },
    "chmod" : function(params) {
      var action_obj = params.data.extra_param;
      OpenNebulaAction.simple_action(params, RESOURCE, "chmod", action_obj);
    },
    "publish": function(params) {
      OpenNebulaAction.simple_action(params, RESOURCE, "publish");
    },
    "unpublish": function(params) {
      OpenNebulaAction.simple_action(params, RESOURCE, "unpublish");
    },
    "hold" : function(params) {
      var action_obj = params.data.extra_param;
      OpenNebulaAction.simple_action(params, RESOURCE, "hold", action_obj);
    },
    "release" : function(params) {
      var action_obj = params.data.extra_param;
      OpenNebulaAction.simple_action(params, RESOURCE, "release", action_obj);
    },
    "add_ar" : function(params) {
      var action_obj = {"ar_template" : params.data.extra_param};
      OpenNebulaAction.simple_action(params, RESOURCE, "add_ar", action_obj);
    },
    "rm_ar" : function(params) {
      var action_obj = params.data.extra_param;
      OpenNebulaAction.simple_action(params, RESOURCE, "rm_ar", action_obj);
    },
    "update_ar": function(params) {
      var action_obj = {"ar_template" : params.data.extra_param};
      OpenNebulaAction.simple_action(params, RESOURCE, "update_ar", action_obj);
    },
    "reserve": function(params) {
      var action_obj = params.data.extra_param;
      OpenNebulaAction.simple_action(params, RESOURCE, "reserve", action_obj);
    },
    "update": function(params) {
      var action_obj = {"template_raw" : params.data.extra_param};
      OpenNebulaAction.simple_action(params, RESOURCE, "update", action_obj);
    },
    "append": function(params) {
      var action_obj = {"template_raw" : params.data.extra_param, append : true};
      OpenNebulaAction.simple_action(params, RESOURCE, "update", action_obj);
    },
    "rename" : function(params) {
      var action_obj = params.data.extra_param;
      OpenNebulaAction.simple_action(params, RESOURCE, "rename", action_obj);
    },
    "getName": function(id){
      return OpenNebulaAction.getName(id, RESOURCE);
    },
    "lock" : function(params) {
      OpenNebulaAction.lock(params, RESOURCE);
    },
    "unlock" : function(params) {
      OpenNebulaAction.simple_action(params, RESOURCE, "unlock");
    },
    "recover" : function(params) {
      var action_obj = {"result": params.data.extra_param};
      OpenNebulaAction.simple_action(params, RESOURCE, "recover", action_obj);
    },
    "stateStr": function(stateId) {
      return STATES_STR[stateId];
    },
    "STATES": STATES
  }

  return Network;
})

File: src/oneprovision/lib/provision/resources/physical/network.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #

require 'provision/resources/physical/physical_resource'

module OneProvision

    # Network
    class Network < PhysicalResource

        # Class constructor
        #
        # @param provider   [Provider] Network provider
        # @param p_template [Hash]     Resource information in hash form
        def initialize(provider, p_template = nil)
            super

            @pool = OpenNebula::VirtualNetworkPool.new(@client)
            @type = 'network'
        end

        # Creates the object in OpenNebula
        #
        # @param cluster_id [Integer] Cluster ID
        #
        # @return [Integer] Resource ID
        def create(cluster_id)
            if @p_template['provision']['count'] && @p_template['ar']
                (Integer(@p_template['provision']['count']) - 1).times do
                    @p_template['ar'] << @p_template['ar'][0]
                end
            end

            net_id = super

            vnet = OpenNebula::VirtualNetwork.new_with_id(net_id, @client)

            require 'opennebula/wait_ext'

            vnet.extend(OpenNebula::WaitExt)

            rc = vnet.wait('READY', 300, 2)

            if !rc && vnet.state_str == 'ERROR'
                vnet.delete
                Utils.exception(OpenNebula::Error.new('Error creating network'))
            elsif !rc
                OneProvisionLogger.warn('Network not READY, '\
                    'check status after provision completes')
            end

            net_id
        end

        # Info an specific object
        #
        # @param id [String] Object ID
        def info(id)
            @one = OpenNebula::VirtualNetwork.new_with_id(id, @client)
            @one.info(true)
        end

        # Destroy network in provider
        #
        # @param provision [OpenNebula::Provision] Provision information
        # @param tf [Hash] Terraform configuration
        def destroy(provision, tf)
            Terraform.p_load

            terraform = Terraform.singleton(@provider, tf)
            terraform.destroy_network(provision, @one.id)
        end

        private

        # Create new object
        def new_object
            @one = OpenNebula::VirtualNetwork.new(
                OpenNebula::VirtualNetwork.build_xml,
                @client
            )
        end

    end

end
